### ISSUE TITLE
`unnecessary_lazy_evaluations` suggestions cause type inferrence errors

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust-clippy#6240]
I tried this code (https://github.com/rust-lang/rust/blob/0da6d42f297642a60f2640ec313b879b376b9ad8/compiler/rustc_trait_selection/src/traits/const_evaluatable.rs#L539):

```rust
    (|| {
        if let Some(a) = AbstractConst::new(tcx, a, a_substs)? {
            if let Some(b) = AbstractConst::new(tcx, b, b_substs)? {
                return Ok(try_unify(tcx, a, b));
            }
        }

        Ok(false)
    })()
    .unwrap_or_else(|ErrorReported| true)
```

I expected to see this happen: Not entirely sure ... maybe the lint shouldn't fire at all here, since there's an explicit type annotation for a unit struct?

Instead, this happened: Since `ErrorReported` is helping inferrence, inference breaks once it's gone.

```
error[E0282]: type annotations needed
   --> compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:539:26
    |
539 |         if let Some(a) = AbstractConst::new(tcx, a, a_substs)? {
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type

warning: unnecessary closure used to substitute value for `Result::Err`
   --> compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:538:5
    |
538 | /     (|| {
539 | |         if let Some(a) = AbstractConst::new(tcx, a, a_substs)? {
540 | |             if let Some(b) = AbstractConst::new(tcx, b, b_substs)? {
541 | |                 return Ok(try_unify(tcx, a, b));
...   |
546 | |     })()
547 | |     .unwrap_or_else(|ErrorReported| true)
    | |_________________________________________^
    |
    = note: `#[warn(clippy::unnecessary_lazy_evaluations)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations
help: Use `unwrap_or` instead
    |
538 |     (|| {
539 |         if let Some(a) = AbstractConst::new(tcx, a, a_substs)? {
540 |             if let Some(b) = AbstractConst::new(tcx, b, b_substs)? {
541 |                 return Ok(try_unify(tcx, a, b));
542 |             }
543 |         }
```

### Meta

- `cargo clippy -V`: clippy 0.0.212 (ffa2e7a 2020-10-24)
